src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -318,13 +542,86 @@ def _prepare_inits(self) -> Set[str]:
+    """
+    # If the dependencies are very large, this flag is indicated to be useful in RAM-constrained
+    # environments. This will be set in both the PEX-INFO and the IPEX-INFO in the produced .ipex
+    # file.
+    self._builder.info.always_write_cache = True
+
+    # PTEX-INFO: Contains resolver settings and source files to copy over into the dehydrated .ipex
+    #            file.
+    chroot = self._builder.chroot()
+    code = [
+      f for f in chroot.get('source') | chroot.get('resource')
+      if f not in ['__main__.py', PexInfo.PATH]
+    ]
+
+    python_setup = self._python_setup_subsystem
+    python_repos = self._python_repos_subsystem
+
+    resolver_settings = dict(
+      indexes=python_repos.indexes,
+      find_links=list(self._all_find_links),
+      allow_prereleases=bool(python_setup.resolver_allow_prereleases),
+    )
+
+    ptex_info = dict(
+      code=code,
+      resolver_settings=resolver_settings,
+      transitive_requirements_by_platform={
+        (
+          platform if platform != 'current' else Platform.current().platform
+        ): reqs
+        for platform, reqs in self._all_transitive_requirements_by_platform.items()
+      },
+    )
+    with temporary_file(permissions=0o644) as ptex_info_file:
+      ptex_info_file.write(json.dumps(ptex_info).encode())
+      ptex_info_file.flush()
+      self._builder.add_resource(filename=ptex_info_file.name, env_filename='PTEX-INFO')
+
+    # IPEX-INFO: The original PEX-INFO, which should be the PEX-INFO in the hydrated .pex file that
+    #            is generated when the .ipex is first executed.
+    ipex_info = self._builder.info.copy()
+    with temporary_file(permissions=0o644) as ipex_info_file:
+      ipex_info_file.write(ipex_info.dump().encode())
+      ipex_info_file.flush()
+      self._builder.add_resource(filename=ipex_info_file.name, env_filename='IPEX-INFO')
+
+    # _ptex_launcher.py: The special bootstrap script to hydrate the .ipex with the fully resolved
+    #                    requirements when it is first executed.
+    ptex_launcher = _IPEX_PREAMBLE
+    with temporary_file(permissions=0o644) as ptex_launcher_file:
+      ptex_launcher_file.write(ptex_launcher.encode())
+      ptex_launcher_file.flush()
+      self._builder.add_source(filename=ptex_launcher_file.name, env_filename='_ptex_launcher.py')
+    # Our .ipex file will use our special preamble.
+    self.set_entry_point('_ptex_launcher')
+
+    # There is an assertion checking this in _ptex_launcher.py due to the use of .zip_unsafe_cache.
+    self._builder.info.zip_safe = False
+    # The PEX-INFO we generate shouldn't have any requirements, or they will fail to bootstrap
+    # because they were unable to find those distributions. Instead, the .pex file produced when the
+    # .ipex is first executed will read and resolve all those requirements from the IPEX-INFO.
+    self._builder.requirements = []
+    self._builder.info._requirements = set()
```

Comment:
You should be cringing each time you have to reach into a private member like this. There should at least be a comment calling the shot. Why do you think you have to do this? Why not give the PTEX a fresh builder?

Reply:
There was some cringing! I'll remember! Giving the ptex a fresh builder sounds like a great idea.

Reply:
Fixed! Thank you for giving me the frame of reference!